### PR TITLE
Refactor along-step for extensibility

### PR DESCRIPTION
**Pull request by sethrj**
**This is currently draft because something's still broken in the Stepper: I'm going to write a nice "along-step" test that will let us probe a single along-step result**

The goal is to allow different along-step kernels based on the physics needs. There is now a generic "along-step" algorithm (function templated on other function arguments) that takes a combination of:
- multiple scattering (so we can create a specialized kernel without it, increasing occupancy)
- propagator (to allow different magnetic field types with arbitrary parameter data)
- energy loss applier (to allow for no-energy-loss neutral particles or faster simulation with eloss disabled by the user)

The along-step kernel has moved out of the physics class and is now simply an action. Further actions can be added. There are two example kernels in place:
- neutral (for photons: no magnetic field, no energy loss)
- along-step general (the current method basically: optional eloss, optional urban msc)

We can/should add more for "no fluctuations" and "uniform magnetic field" (for testing) but I wanted to wait for feedback on the basic interfaces.

Current deficiencies:
- The along-step classes have no easy way of being auto generated because they rely on different data and different on-device operators that use the data.
- I could move the UrbanMsc data out of the physics like I did with fluctuation, but it would require changing the importdata and such so I'm going to postpone that

### CODE REVIEW

**Comment by amandalund** [comment]
@sethrj I think it's broken because we're not actually registering the along-step action (so never launching the along-step kernel).

**Comment by sethrj** [comment]
@amandalund Bingo! Even though I added its "action ID", because I never inserted it the action ID was being reused for something else 😱 

Pushing a fix now!

File: src/celeritas/global/alongstep/AlongStep.hh [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    {
+        msc.apply_step(track, &local);
+    }
+    else
+    {
+        // Step might have changed due to geometry boundary
+        local.step_limit.step = local.geo_step;
+    }
+
+    // Update track's lab-frame time using the beginning-of-step speed
```
Comment (whokion) [inline src/celeritas/global/alongstep/AlongStep.hh:97]: Wondering whether this block should go before msc.apply_step (i.e, time update based on the geometry step, not by true step - I am not sure which one is correct, but Geant4, as far as I know, updates the global time before the msc sampling)? In general, the field integration may be performed by the time interval (instead of the distance traveled) alternatively.

File: src/celeritas/global/alongstep/detail/UrbanMsc.hh [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Apply Urban multiple scattering to a track.
+ */
+class UrbanMsc
```
Comment (whokion) [inline src/celeritas/global/alongstep/detail/UrbanMsc.hh:27]: Wasn't this defined in [the another file](https://github.com/celeritas-project/celeritas/pull/457/files#diff-4828f3e911e6f130b9fa8a108f319437347357861477cf84ceb65825cd6d4a66) or is this duplication under detail namespace intended for different uses?

File: src/celeritas/phys/PhysicsStepUtils.hh [hunk 3]
```diff
@@ -180,35 +156,25 @@ calc_physics_step_limit(const MaterialTrackView& material,
...
                   "Incompatible energy types");
 
     auto ppid = physics.eloss_ppid();
-    if (!ppid)
-    {
-        // No energy loss processes for this particle
-        return Energy{0};
-    }
-
     const real_type pre_step_energy = value_as<Energy>(particle.energy());
```
Comment (whokion) [inline src/celeritas/phys/PhysicsStepUtils.hh:178]: Since this is based on the currently, we do not have the approximation below the lowest kinetic energy (i.e, just returning eloss = particle.energy()).  Will it be fixed up later or are you going to actually drop it or move it somewhere else?

**Review by whokion — DISMISSED** [review]
This is a lot of code changes!  Added the first set of review, but have not tested the branch on my platform yet (which I may do it later).

File: src/celeritas/field/PropagatorLauncher.hh [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/field/PropagatorLauncher.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+namespace celeritas
+{
+
+struct UrbanMsc
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/celeritas/field/PropagatorLauncher.hh:13]: Oh goodness, I left this file in there by mistake. Thanks.

File: src/celeritas/field/PropagatorLauncher.hh [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        auto msc_result = msc_scatter(rng);
+
+        // Update full path length traveled along the step based on MSC to
+        // correctly calculate energy loss, step time, etc.
+        CELER_ASSERT(geo_step <= msc_result.step_length
+                     && msc_result.step_length <= local->step_limit.step);
+        local->step_limit.step = msc_result.step_length;
+
+        // Update direction and position
+        geo.set_dir(msc_result.direction);
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/celeritas/field/PropagatorLauncher.hh:87]: Yep, zombie file :(

File: src/celeritas/field/PropagatorLauncher.hh [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+            track.make_material_view().material_id(),
+            sim.num_steps() == 0,
+            local->step_limit.step);
+
+        auto rng             = track.make_rng_engine();
+        auto msc_step_result = msc_step_limit(rng);
+        track.make_physics_step_view().msc_step(msc_step_result);
+
+        // Use "straight line" path calculated for geometry step
+        ? local->geo_step = msc_step_result.geom_path;
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/celeritas/field/PropagatorLauncher.hh:52]: whoops, zombie file :(

File: src/celeritas/global/alongstep/detail/UrbanMsc.hh (hunk 2 above)
Comment (sethrj) [reply src/celeritas/global/alongstep/detail/UrbanMsc.hh:27]: Yep the other one was a zombie file :(

File: src/celeritas/global/alongstep/AlongStep.hh (hunk 1 above)
Comment (amandalund) [reply src/celeritas/global/alongstep/AlongStep.hh:97]: Section 3.2.4 of the PRM says the time change is calculated using the true path length -- but it also says it's calculated from both the initial and final particle velocity, and as far as I could tell in the code only the initial velocity is used, so it might be good to double check. I think global time is updated [here](https://geant4.kek.jp/lxr/source/tracking/src/G4SteppingManager2.cc#L470), and msc is sampled just before in `AlongStepDoIt`?

Comment (whokion) [reply src/celeritas/global/alongstep/AlongStep.hh:97]: The sequence of Geant4 AlongStepDoIt goes as Transporation -> msc -> eIoni and the actual place for calculating time in Geant4 is inside G4Transporation::AlongStepDoIt (see this [line](https://geant4.kek.jp/lxr/source/processes/transportation/src/G4Transportation.cc#L529) and update time information based on that).  So, it seems that there is, at least, inconsistency between PRM and the actual time used in the code which is calculated with geometry_step/initialVelocity - in this case, the average time is equivalent to the initial time as there is no change in energy (if there is no electric field).  To me, if the average velocity and the true path should be used, then the time update should be done even after eloni as the energy (so momentum) is changing during the eloss, but not during msc (even though b.

**Review by amandalund — COMMENTED** [review]
Looking great @sethrj! Nothing major from me. I ran the demo loop on simple cms and testem3 with this branch and it looked like everything was working as expected.

File: src/celeritas/global/alongstep/AlongStep.hh (hunk 1 above)
Comment (amandalund) [reply src/celeritas/global/alongstep/AlongStep.hh:97]: ahh I see, thanks @whokion -- good catch!

**Comment by sethrj** [comment]
Awesome! One of the hacky bits was the along step local state: do you think it could be improved with encapsulation rather than just being an open struct?

File: src/CMakeLists.txt [hunk 7]
```diff
@@ -261,9 +269,6 @@ celeritas_strip_alias(_TARGET celeritas)
 add_library(Celeritas::Core ALIAS ${_TARGET})
 add_library(Celeritas::celeritas ALIAS ${_TARGET})
 
-# Require at least C++14
-set_property(TARGET ${_TARGET} PROPERTY CXX_STANDARD 14)
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/CMakeLists.txt:265]: Nope 😒 thought I caught all these bad merge issues

**Comment by amandalund** [comment]
> Awesome! One of the hacky bits was the along step local state: do you think it could be improved with encapsulation rather than just being an open struct?

Slightly clumsy I suppose, but I wasn't offended by it :)

**Comment by sethrj** [comment]
This afternoon I'll update this with @whokion 's msc interface changes. I'll also try adding field support, but for now it may require the field propagation working for gammas...

File: src/celeritas/field/PropagatorLauncher.hh [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    {
+        if (step <= this->urban_data.params.geom_limit)
+            return false;
+
+        auto phys = track.make_physics_view();
+        if (!phys.msc_ppid())
+            return false;
+
+        auto particle = track.make_particle_view();
+        return particle.energy() > this->urban_data.params.energy_limit;
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/celeritas/field/PropagatorLauncher.hh]: No, the process id doesn't do any energy discrimination: just particle type. Although this file is dead, the `UrbanMsc.hh` file has something similar, but this should be exactly identical.

File: src/celeritas/global/alongstep/AlongStep.hh [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+    // TODO: scope the 'views' so that the lifetimes don't overlap between this
+    // function and helper class functions
+    auto sim = track.make_sim_view();
+
+    // True step is the actual path length traveled by the particle, including
+    // within-step MSC
+    AlongStepLocalState local;
+    local.step_limit = sim.step_limit();
+    CELER_ASSERT(local.step_limit);
+    if (local.step_limit.step == 0)
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/celeritas/global/alongstep/AlongStep.hh:53]: This is also exactly what we're doing now, so I'll add a checkbox on #459 to revisit this.

File: src/celeritas/global/alongstep/AlongStep.hh [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+        return;
+    }
+
+    local.geo_step = local.step_limit.step;
+    bool use_msc   = msc.is_applicable(track, local.geo_step);
+    if (use_msc)
+    {
+        msc.calc_step(track, &local);
+        CELER_ASSERT(local.geo_step > 0);
+        CELER_ASSERT(local.step_limit.step >= local.geo_step);
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/celeritas/global/alongstep/AlongStep.hh:72]: Yes, the `step` here is the true step.

File: src/celeritas/global/alongstep/AlongStep.hh (hunk 1 above)
Comment (sethrj) [reply src/celeritas/global/alongstep/AlongStep.hh:97]: It seems to me we should be using the true path for the update, and I would imagine the average speed would be more accurate. Again, this implementation is equivalent to what's on master, so I'm adding a note on #459 and leaving it as-is for now.

File: src/celeritas/phys/PhysicsStepUtils.hh (hunk 3 above)
Comment (sethrj) [reply src/celeritas/phys/PhysicsStepUtils.hh:178]: Yep, when I merged master yesterday (after merging your work) this should be fixed.

File: src/celeritas/phys/PhysicsData.hh [hunk 11]
```diff
@@ -474,10 +461,7 @@ inline void resize(PhysicsStateData<Ownership::value, M>* state,
     CELER_EXPECT(size > 0);
     CELER_EXPECT(params.scalars.max_particle_processes > 0);
     resize(&state->state, size);
-    if (params.hardwired.msc)
-    {
-        resize(&state->msc_step, size);
-    }
+    resize(&state->msc_step, size);
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/celeritas/phys/PhysicsData.hh:468]: No, not that I could think of. As you say, I don't think it should be a big deal. We can potentially move the `msc_step` stuff into the along-step `Action` as state data, but that'll be a task for somewhere down the road.

**Comment by sethrj** [comment]
All right, I think this is ready: it should just be refactoring; the new field along-step I'll submit in a new PR. Some of the discussion points @whokion brought up I've moved over to #459: now that we have along-step and stepper tests we can actually play around with the algorithms and add tests :)

**Comment by whokion** [comment]
@sethrj Please ignore those compilation errors - sorry that I did not checkout the branch.

File: src/celeritas/global/alongstep/AlongStep.hh (hunk 9 above)
[the remark this answers is not in the document]
Comment (amandalund) [reply src/celeritas/global/alongstep/AlongStep.hh:54]: We'll need to remove the assertion that `step > 0` since we _can_ still have zero step length here. As @whokion said, Geant4 invokes at-rest processes at the beginning of the step, while we currently do it at the end, along with the rest of the discrete actions. If a positron loses all energy over the step, it _might_ annihilate at the end of the step, or the discrete action might be rejected in the integral method (since the process is sampled based on pre-step macro xs, so e.g. it might sample bremsstrahlung and then reject the interaction). If that happens, the positron will take one more step and the step length will be zero at this point. I think this might also be why we see a difference in the number of steps per track between Celeritas and Geant4 for positrons:
steps_per_track_testem3_zoom

**Comment by whokion** [comment]
celeritas_global_Stepper fails at
AlongStep.hh:53:
celeritas: internal assertion failed: local.step_limit.step > 0

File: src/celeritas/global/alongstep/AlongStep.hh (hunk 9 above)
[the remark this answers is not in the document]
Comment (sethrj) [reply src/celeritas/global/alongstep/AlongStep.hh:53]: Aha, good point @amandalund .

File: src/celeritas/global/alongstep/detail/EnergyLossFluctApplier.hh [hunk 12]
```diff
@@ -68,48 +68,56 @@ EnergyLossFluctApplier::operator()(CoreTrackView const& track,
...
     auto particle = track.make_particle_view();
     if (particle.energy() < phys.scalars().eloss_calc_limit)
     {
         // Immediately stop low-energy particles
-        auto step = track.make_physics_step_view();
-        step.deposit_energy(particle.energy());
-        particle.energy(zero_quantity());
-        return;
+        // TODO: this should happen before creating tracks from secondaries
+        // *OR* after slowing down tracks: duplicated in
```
Comment (amandalund) [inline src/celeritas/global/alongstep/detail/EnergyLossFluctApplier.hh:84]: Why is this?

Comment (sethrj) [reply src/celeritas/global/alongstep/detail/EnergyLossFluctApplier.hh:84]: It seemed to me that applying the cut after the geometry step and before doing the next energy loss step makes it more likely that we'll kill tracks on the boundary, and we're doing less work per step. If we apply the cut at the end of the step (like we originally did with our cutoff params, even later than my comment here suggests), we're more likely to kill because of a discrete interaction or slowing down.

Comment (whokion) [reply src/celeritas/global/alongstep/detail/EnergyLossFluctApplier.hh:84]: Not sure whether I am on the same page, but we should clearly distinguish the tracking cut vs. the production cut (on secondaries from discrete interactions).

**Review by amandalund — APPROVED** [review]
Looks good to me @sethrj!

**Comment by whokion** [comment]
Just for record, this branch still hits an assertion from demo-loop with testem3-flat.gdml and testem3.100k.hepmc3 on NVIDIA GeForce RTX 3090,
SBEnergyDistribution.hh:168: celeritas: internal assertion failed: xs >= 0 && xs <= 1 / inv_max_xs_  
but passes the same test on V100.

**Comment by amandalund** [comment]
@whokion I added it to the list in #459 along with the failing assertion from a geo-limited positron with zero mfp that I'm still seeing.

**Comment by sethrj** [comment]
@whokion Is this assertion not there on previous master? I thought that none of the high-level tests changed results, but I guess the coverage is incomplete...

**Comment by whokion** [comment]
@sethrj The assertion was in the previous master too (only on my local RTX3090, not seen on V100 which seems very odd to me).  One strange symptom is that many bremsstrahlung photons blow the cutoff of lAr (0.00619986) of testem3-flat geometry are produced by e+ which was questioned by @amandalund and is not seen on V100.

**Comment by sethrj** [comment]
Interesting. Are they running with the same number of threads? If it's not an error introduced by this PR we should track it in a sea rate issue (or the big "MSC" one)

**Comment by whokion** [comment]
@sethrj @amandalund Using the testem3-flat-gpu.root produced on wc.fnal.gov (V100), demo-loop with testem3 actually runs without any problem on opticks.fnal.gov (RTX3090) too - strange enough! Differences that I see are the size of the root file
```
-rw-r--r--  1 syjun gpudev    49481 Jul 22 13:20 testem3-flat-gpu.root.build_on_opticks
-rw-r--r--  1 syjun gpudev    49958 Jul 25 14:30 testem3-flat-gpu.root_build_on_wc
```
which were produced with the same celeritas code, but with different versions of Geant4 (so SB related data)
```
Geant4-11.0.0/data/G4EMLOW8.0/brem_SB (opticks)
Geant4-11.0.1/data/G4EMLOW8.1/brem_SB (wc)
```
As it seems that the issue is not related to celeritas code itself, I probably stop to trace out the source of the problem with RTX3090 further with the Geant4 version 11.0.0.

**Comment by sethrj** [comment]
It could still be celeritas, but just an "unusual" thing that we happen to hit with one set of data but not the other.

**Comment by amandalund** [comment]
@whokion do you hit the assertion on the V100 using the root file created on the RTX3090?

**Comment by whokion** [comment]
@amandalund A good cross check!  Yes, I can reproduce the same assertion with testem3-flat-gpu.root which was created on opticks (RTX3090).
